### PR TITLE
terraform_0_{12,13,14,15}: help users upgrade

### PR DESCRIPTION
**Pull request by grahamc**
cc @zimbatm 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by grahamc** [comment]
I wonder if we should restore 0.13 temporarily so people don't need to go to a different channel for the upgrade? Also I wonder if we can start communicating this change on the current stable.

**Comment by endocrimes** [comment]
I'd be a proponent of restoring 0.13 for a while, otherwise new error messages LGTM

File: pkgs/top-level/aliases.nix [hunk 1]
```diff
@@ -818,10 +818,19 @@ mapAliases ({
   telepathy_salut = telepathy-salut; # added 2018-02-25
   telnet = inetutils; # added 2018-05-15
   terraform_1_0_0 = terraform_1_0; # added 2021-06-15
-  terraform_0_12 = throw "terraform 0.12 has been removed"; # added 2021-06-15
-  terraform_0_13 = throw "terraform 0.13 has been removed"; # added 2021-06-15
-  terraform_0_14 = throw "terraform 0.14 has been removed"; # added 2021-06-15
-  terraform_0_15 = throw "terraform 0.15 has been removed"; # added 2021-06-15
+  terraform_0_12 = throw ''
```
Comment (andir) [inline pkgs/top-level/aliases.nix:821]: Why do we have two different flavors of throws here? Why shouldn't a user of 0.14 receive the nice error message?

Comment (grahamc) [reply pkgs/top-level/aliases.nix:821]: I thought they were all nice... 0.14 users don't need to go by way of 0.13, so I didn't think needed the more in-depth explanation of how to go from 0.12 to 0.13.

**Comment by endocrimes** [comment]
@grahamc We also probably need to add a release note for the original PR about the removals of all pre-1.0 terraform versions tbh

**Comment by grahamc** [comment]
> @grahamc We also probably need to add a release note for the original PR about the removals of all pre-1.0 terraform versions tbh

https://github.com/NixOS/nixpkgs/pull/126919#issuecomment-863204632

:thumbsup: agreed

**Comment by grahamc** [comment]
To make matters more significant, `terraform` was `terraform_0_12` making it very likely a lot of users are waking up to a very significant bump in their tf.

**Comment by grahamc** [comment]
I think we essentially need to bring back terraform versions 0.13, 0.14, 0.15, and 1.0 -- then drop 0.14 then 0.15, and maybe hold on to 0.13 until after 21.10 releases.

**Comment by maxeaubrey** [comment]
Changing the default version of `terraform` at the same time as dropping 0.12 doesn't feel great. Changing the default and leaving it in for at least another month would be IMO a little nicer to end users.

At least my position is that changing the default version and dropping the previous default at the same time is a bit forceful, especially for software like Terraform.

**Comment by zimbatm** [comment]
Thanks, that's a well-worded PR. For now I propose to just revert the deletion: https://github.com/NixOS/nixpkgs/pull/127239

**Comment by endocrimes** [comment]
Closing because #127239 got merged - We should probably use this when we do eventually remove those though.

**Pull request closed without merging**